### PR TITLE
Transparent SAML error Response for MFA entities

### PR DESCRIPTION
**Pull request by MKodde**
When a SP request a AuthnContext and transparent_authn_context is enabled, we pass on to the IDP. There is no way for us to know the meaning or value for this AuthnContext. If a user can't use the requested AuthnContext, the SP should be informed so it can decide what to do.

When an SP is configured for transparant_authn_context, an NoAuthContext saml error from the IDP should be copied to the SP, so it can act accordingly.

https://www.pivotaltracker.com/story/show/173602383

### CODE REVIEW

File: library/EngineBlock/Corto/Module/Bindings.php [hunk 1]
```diff
@@ -415,6 +415,18 @@ public function receiveResponse($serviceEntityId, $expectedDestination)
                     )
                 );
 
+                // If this is a transparent MFA authentication, we want to forward the error response
+                $originalRequest = $this->_server->findRequestFromRequestId($sspResponse->getInResponseTo());
```
Comment (Badlapje) [inline library/EngineBlock/Corto/Module/Bindings.php:419]: shouldn't this variable be named `$spResponse`?  If not, what does the extra s stand for?

File: tests/unit/OpenConext/EngineBlock/Service/MfaHelperTest.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        self::assertFalse($isTransparent);
+    }
+
+    private function createIdP($spEntityId, $isSpInMfaEntities, $isTransparent)
+    {
+        $idp = m::mock(IdentityProvider::class);
+        $mfaEntities = m::mock(MfaEntityCollection::class);
+        $mfaEntity = null;
+        if ($isSpInMfaEntities) {
+            if ($isTransparent) {
```
Comment (Badlapje) [inline tests/unit/OpenConext/EngineBlock/Service/MfaHelperTest.php:105]: can be turned into a ternary operator, but i'm fine with it as is as well.

**Review by Badlapje — APPROVED** [review]
1 thing i'm not clear on and one nitpick (up to you whether it makes sense).  Code-wise this looks good to me!

File: library/EngineBlock/Corto/Module/Bindings.php (hunk 1 above)
Comment (MKodde) [reply library/EngineBlock/Corto/Module/Bindings.php:419]: ssp == simple saml php. It denotates we got this from the SSP library.

File: library/EngineBlock/Corto/Module/Service/AssertionConsumer.php [hunk 3]
```diff
@@ -90,6 +90,13 @@ public function serve($serviceName, Request $httpRequest)
         $application = EngineBlock_ApplicationSingleton::getInstance();
         $log = $application->getLogInstance();
 
+        if ($receivedResponse->isTransparentErrorResponse()) {
+            $log->info('Response contains NoPassive status code: responding with NoPassive status to SP');
```
Comment (MKodde) [inline library/EngineBlock/Corto/Module/Service/AssertionConsumer.php:94]: Ouch, forgot to update the log entry :blush:

File: library/EngineBlock/Corto/ProxyServer.php [hunk 4]
```diff
@@ -868,7 +878,11 @@ public function getReceivedRequestFromResponse(EngineBlock_Saml2_ResponseAnnotat
                 EngineBlock_Exception::CODE_NOTICE
             );
         }
+        return $this->findRequestFromRequestId($requestId);
+    }
 
+    public function findRequestFromRequestId(string $requestId): ?EngineBlock_Saml2_AuthnRequestAnnotationDecorator
+    {
```
Comment (MKodde) [inline library/EngineBlock/Corto/ProxyServer.php:885]: I dumbed down the original method to allow for retrieving the Request from a string. This allows getting the Request from the SSP library received 'inResponseTo' string.

**Review by phavekes — APPROVED** [review]

**Pull request merged**